### PR TITLE
test runner rework

### PR DESCRIPTION
**Pull request by tavip**
Rework the test runner from a Makefile based one to a python script
that generates clean test output summary similar with the Linux
kernel build output. This makes it easier to understand what tests
have failed and why when run from the command line.

The runner can also generate junit xml files that we can integrate
with the CI to generate test reports per PR.

This PR also reworks the build system by moving the autoconfiguration bits to a separate Makefile and by generating persistent host configuration: 

- Makefile.conf: stores system configuration like CFLAGS, LDFLAGS, LDLIBS, etc.
- include/autoconf.h: header file that allows build time identification if certain features are available
- tests/autoconf.sh: shell variables that allows the test runner to identify if certain features are available

---
This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/400)

### CODE REVIEW

File: tools/lkl/tests/net-test.c [hunk 1]
```diff
@@ -44,41 +77,59 @@ in_cksum(const u_short *addr, register int len, u_short csum)
...
+	lkl_test_logf("pinging %s\n", inet_ntoa(saddr.sin_addr));
+
+	sock = lkl_sys_socket(LKL_AF_INET, LKL_SOCK_RAW, LKL_IPPROTO_ICMP);
 	if (sock < 0) {
-		snprintf(str, len, "socket error (%s)", lkl_strerror(sock));
+		lkl_test_logf("socket error (%s)\n", lkl_strerror(sock));
 		return TEST_FAILURE;
 	}
 
 	icmp = malloc(sizeof(struct icmphdr *));
```
Comment (thehajime) [inline tools/lkl/tests/net-test.c]: maybe it's okay but `sizeof(struct lkl_icmphdr *)` instead ?

File: tools/lkl/Makefile [hunk 2]
```diff
@@ -2,192 +2,89 @@
...
-else
-  $(error Unrecognized platform: $(OUTPUT_FORMAT))
 endif
 
 ifneq ($(OUTPUT),)
-OUTPUT := $(OUTPUT)/tools/lkl/
+  OUTPUT := $(OUTPUT)/tools/lkl/
 else
-OUTPUT := $(CURDIR)/
+ OUTPUT := $(CURDIR)/
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/Makefile]: Good catch, I will fix it.

File: tools/lkl/tests/boot.c [hunk 3]
```diff
@@ -598,84 +310,31 @@ static int test_getdents64(char *str, int len)
...
 		return TEST_FAILURE;
 
 	for (pos = buf; pos - buf < ret; pos += de->d_reclen) {
 		de = (struct lkl_linux_dirent64 *)pos;
 
-		wr += snprintf(str + wr, len - wr, "%s ", de->d_name);
-		if (wr >= len)
-			break;
+		wr += lkl_test_logf("%s ", de->d_name);
+		if (wr >= 70) {
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/tests/boot.c]: Yes, good idea, will fix it.

File: tools/lkl/tests/boot.c [hunk 4]
```diff
@@ -756,34 +409,34 @@ static int test_syscall_thread(char *str, int len)
...
 	}
 
 	ret = lkl_sys_write(pipe_fds[1], tmp, sizeof(tmp));
 	if (ret != sizeof(tmp)) {
 		if (ret < 0)
-			snprintf(str, len, "write: %s", lkl_strerror(ret));
+			lkl_test_logf("write error: %s\n", lkl_strerror(ret));
 		else
-			snprintf(str, len, "write: short write: %ld", ret);
+			lkl_test_logf("short write: %ld", ret);
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/tests/boot.c]: Nice catch, will fix it.

File: tools/lkl/tests/boot.c [hunk 5]
```diff
@@ -715,15 +372,13 @@ static int test_semaphore(char *str, int len)
 	lkl_host_ops.sem_up(sem);
 	lkl_host_ops.sem_free(sem);
 
-	snprintf(str, len, "%ld", ret);
-
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/tests/boot.c]: There is not useful information to print, this tests always printed TEST_SUCCESS (and the tests status is printed anyway).

File: tools/lkl/tests/hijack-test.sh [hunk 6]
```diff
@@ -1,591 +1,741 @@
...
-(echo "$ans" | grep "100756k") || true
-fi
+    # Make sure our device has the addresses we expect
+    addr=$(${hijack_cfg} ip addr)
+    echo "$addr" | grep eth0
+    echo "$addr" | grep $(ip_lkl)
+    echo "$addr" | grep "$TEST_MAC0"
+    echo "$addr" | grep "$(ip6_lkl)"
+    ! echo "$addr" | grep "WARN: failed to free"
+}
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/tests/hijack-test.sh]: yes, the new command line tests is platform independent and we should not have any critical message printed to screw it up.

File: tools/lkl/tests/net-test.c (hunk 1 above)
Comment (tavip) [reply tools/lkl/tests/net-test.c]: Yes, good catch, I'll fix it.

File: tools/lkl/tests/hijack-test.sh [hunk 7]
```diff
@@ -1,591 +1,741 @@
...
+    ${netperf} $(ip_host) TCP_MAERTS
 }
+
+
+test_tap_qdisc()
+{
+    set -e
+
+    if [ -n "$LKL_HOST_CONFIG_ANDROID" ]; then
+        return $TEST_SKIP
```
[the remark this answers is not in the document]
Comment (tavip) [reply tools/lkl/tests/hijack-test.sh]: It looks like Android does not support the standard qdiscs, here is the error I am getting:

+ LKL_HIJACK_CONFIG_FILE=/data/local/tmp/lkl/hijack-test.conf /data/local/tmp/lkl/bin/lkl-hijack.sh tc -s -d qdisc show
Android does not support qdisc 'noqueue'
Android does not support qdisc 'fq'
qdisc noqueue 0: dev lo root refcnt 2
Sent 0 bytes 0 pkts (dropped 0, overlimits 0)
qdisc fq 8001: dev eth0 root refcnt 2 [cannot parse qdisc parameters]
Sent 0 bytes 0 pkts (dropped 0, overlimits 0)

We can still probably fix this, but it needs more investigation.

File: tools/lkl/tests/boot.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (thehajime) [reply tools/lkl/tests/boot.c]: okay

File: tools/lkl/tests/hijack-test.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (thehajime) [reply tools/lkl/tests/hijack-test.sh]: okay, agree.

File: tools/lkl/tests/test.sh [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        else
+            SU=""
+        fi
+        WRAPPER="adb shell $SU"
+    fi
+
+    echo "$@" | $WRAPPER sh -x
+}
+
+adb_push()
```
Comment (thehajime) [inline tools/lkl/tests/test.sh]: maybe we can remove/delete tools/lkl/tests/android-test.sh ?

File: tools/lkl/tests/hijack-test.sh [hunk 9]
```diff
@@ -1,591 +1,741 @@
...
-if [ -z `which mkfifo` ]; then
-    echo "WARNIG: no mkfifo command, skipping PIPE tests."
-else
+test_tap_cleanup()
+{
+    tap_cleanup
+    tap_cleanup 1
+}
+
+test_tap_ping_host()
```
Comment (thehajime) [inline tools/lkl/tests/hijack-test.sh]: I don't know how it works on android.
IIRC android doesn't have tap support but the following test results says `TEST       ok       test_tap_ping_host `.  

https://circleci.com/gh/tavip/linux/134?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: tools/lkl/tests/test.sh (hunk 8 above)
Comment (thehajime) [reply tools/lkl/tests/test.sh]: `make test` could be replaced as well

File: tools/lkl/tests/hijack-test.sh (hunk 9 above)
Comment (tavip) [reply tools/lkl/tests/hijack-test.sh]: It does work, it just needs the right permissions, which in this case are: user=vpn,group=vpn and additional group=net_admin. See the tap_prepare and tap_setup in net-setup.sh

File: tools/lkl/tests/test.sh (hunk 8 above)
Comment (tavip) [reply tools/lkl/tests/test.sh]: Yes, these are removed in the last patch of the series.

**Comment by tavip** [comment]
The checkpatch error is about using __attributte__(printf) instead of the __printf define in the kernel headers. There is a definition in tools/include but including those headers pulls in other dependincies that are failing on some platfors, so I think it is sensible for now to just go with __attribute__.

File: tools/lkl/tests/net.sh [hunk 10]
```diff
@@ -1,102 +1,175 @@
...
+        echo "not a posix environment"
+        return $TEST_SKIP
+    fi
 
-    ./net-test tap lkl_ptt1 192.168.14.1 192.168.14.2 24
+    case "$1" in
+    "loopback")
+        ;;
+    "pipe")
+        if [ -z $(lkl_test_cmd which mkfifo) ]; then
```
Comment (thehajime) [inline tools/lkl/tests/net.sh]: do you see why x86_64 test skips pipe test ?

https://circleci.com/gh/tavip/linux/151?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: tools/lkl/tests/hijack-test.sh (hunk 9 above)
Comment (thehajime) [reply tools/lkl/tests/hijack-test.sh]: that's new to me (cc: @phhusson @linhua55 and #379). Great and thanks !

File: tools/lkl/tests/test.sh (hunk 8 above)
Comment (thehajime) [reply tools/lkl/tests/test.sh]: Thanks,

in tools/lkl/Makefile there is still the target `test valgrind gdb` which calls `make -C tests`.

https://github.com/lkl/linux/pull/400/files#diff-bb2bc012ca931e05448c2b62f99c55abR107

So removing tools/lkl/tests/Makefile lets those targets orphaned while removing android-test.sh is okay.

I think `make gdb/tests/valgrind` is still useful so how about keeping `tools/lkl/tests/Makefile` ?

**Comment by thehajime** [comment]
> The checkpatch error is about using attributte(printf) instead of the __printf define in the kernel headers. There is a definition in tools/include but including those headers pulls in other dependincies that are failing on some platfors, so I think it is sensible for now to just go with attribute.

Okay, agree.

File: tools/lkl/tests/net.sh (hunk 10 above)
Comment (tavip) [reply tools/lkl/tests/net.sh]: There was an issue in the test, I've fixed it and will push a new version soon.

File: tools/lkl/tests/test.sh (hunk 8 above)
Comment (tavip) [reply tools/lkl/tests/test.sh]: For gdb and valgrind, I think we should add an env var to select the proper wrapper in test.sh. That way we can run gdb/valgrind for all commands not only boot. 

As for the tests target, I guess we can add a target into the top Makefile to run tests/run.py.

I'll do these changes in the new version I'll push soon.

File: tools/lkl/tests/net.sh (hunk 10 above)
Comment (thehajime) [reply tools/lkl/tests/net.sh]: thanks, will look at your new version.

File: tools/lkl/tests/test.sh (hunk 8 above)
Comment (thehajime) [reply tools/lkl/tests/test.sh]: thanks too.  I agree with what you're going to update.

**Comment by tavip** [comment]
@thehajime Fixed the pipe test. Also added support for gdb and valgrind by either using --gdb or --valgrind options with the test runner or by setting VALGRIND or GDB env vars to non-null values.

Also added nightly build that runs valgrind on x86_64.

**Comment by thehajime** [comment]
Thanks, valgrind nightly build needs a bit more work but I will create another PR for that.

**Pull request merged**